src/CBLURLEndpointListener.cc

```diff
@@ -0,0 +1,216 @@
+        }
+        return succ;
+    };
+
+    CBLDatabase* cblDb = _conf.collections[0]->database();
+    bool succ = true;
+    cblDb->c4db()->useLocked([&](C4Database* db) {
+        slice dbname = db->getName();
+        if ( (succ = c4listener->shareDB(dbname, db)) ) {
+            for (unsigned i = 0; i < _conf.collectionCount; ++i) {
+                _conf.collections[i]->useLocked([&](C4Collection* coll) {
+                    succ = c4listener->shareCollection(dbname, coll);
+                });
+                if (!succ) break;
+            }
+        }
+    });
+
+    return ret(succ);
+}
+
+void CBLURLEndpointListener::stop() {
+    std::scoped_lock lock(_mutex);
+    if (_c4listener) {
+        delete _c4listener;
+        _c4listener = nullptr;
+    }
+}
+
+std::string CBLURLEndpointListener::dumpConfig() {
+    return {};
+}
+
+CBLTLSIdentity* CBLURLEndpointListener::effectiveTLSIdentity(bool persistent) {
+    if (_conf.disableTLS)
+        return nullptr;
+
+    if (!_effectiveTLSIdentity)
+        _effectiveTLSIdentity = _conf.tlsIdentity ? _conf.tlsIdentity : anonymousTLSIdentity(persistent).get();
+
+    return _effectiveTLSIdentity;
+}
+
+Retained<CBLTLSIdentity> CBLURLEndpointListener::anonymousTLSIdentity(bool persistent) {
+    Retained<CBLTLSIdentity> identity;
+    alloc_slice label;
+    std::unique_ptr<CBLKeyPair, void(*)(CBLKeyPair*)> keypair{
+        nullptr,
+        [](CBLKeyPair* k) {
+            CBLKeyPair_Release(k);
+        }
+    };
+    if (persistent) {
+#if !defined(__linux__) && !defined(__ANDROID__)
+        label = labelForAnonymousTLSIdentity();
+        if (!label) return nullptr;
+
+        identity = CBLTLSIdentity::IdentityWithLabel(label);
+        C4Error error{};
+        if (identity) {
+            CBL_Log(kCBLLogDomainListener, kCBLLogVerbose, "Found anonymous identity by label = '%.*s'", (int)label.size, (char*)label.buf);
+
+            CBLTimestamp expire = identity->expiration();
+            if (expire/1000 > kCBLAnonymousIdentityMinValidTimeAllowed) {
+                return identity;
+            }
+
+            CBL_Log(kCBLLogDomainListener, kCBLLogVerbose, "Delete anonymous identity of label = '%.*s' (expiration = %ld)", (int)label.size, (char*)label.buf, (long)(expire/1000));
+
+            CBLTLSIdentity_DeleteIdentityWithLabel(label, cbl_internal::external(&error));
+            if (error.code) {
+                CBL_Log(kCBLLogDomainListener, kCBLLogWarning, "Error when delete anonymous identity of label = '%.*s', error = %d/%d", (int)label.size, (char*)label.buf, error.domain, error.code);
+            }
+        }
+
+        if (error.code && error.code != kC4ErrorNotFound) {
+            C4Error::raise(error);
+        }
+#else //#if !defined(__linux__) && !defined(__ANDROID__)
+        C4Error::raise(LiteCoreDomain, kC4ErrorUnimplemented, "No persistent key support");
+#endif
+    } else {
+        keypair.reset(CBLKeyPair_GenerateRSAKeyPair(fleece::nullslice, nullptr));
```

Comment:
Use C++ level API instead of public C api.

Reply:
This is a function in Private.h and implemented directly in CBLTLSIdentity_CAPI.cc. Since it's a private function, I did not follow the format rigorously. I can put it in CBLTLSIdentity_Internal.hh